### PR TITLE
Converted login modal UI to react

### PR DESCRIPTION
**Pull request by mzparacha**
## Link to Issue
Closes: https://github.com/hicommonwealth/commonwealth/issues/3702

## Description of Changes
- Converted the login modal from legacy mithril to react

## Test Plan

- Login/logout works inside and outside of community with and without clearing browser storage/cookies
- Use different chain wallets without page reload should work

## Deployment Plan
N/A

## Other Considerations
N/A

### Metamask/Ethereum

- [x] **Inside community**) Login works before clearing localstorage
- [x] **Inside community**) Login works after clearing localstorage
- [x] **Inside community**) Logout works before clearing localstorage
- [x] **Inside community**) Re-login works without clearing localstorage
- [x] **Outside community**) Login works before clearing localstorage
- [x] **Outside community**) Login works after clearing localstorage
- [x] **Outside community**) Logout works before clearing localstorage
- [x] **Outside community**) Re-login works without clearing localstorage

### Keplr (Cosmos Hub/Osmosis)

- [x] **Inside community**) Login works before clearing localstorage
- [x] **Inside community**) Login works after clearing localstorage
- [x] **Inside community**) Logout works before clearing localstorage
- [x] **Inside community**) Re-login works without clearing localstorage
- [x] **Outside community**) Login works before clearing localstorage
- [x] **Outside community**) Login works after clearing localstorage
- [x] **Outside community**) Logout works before clearing localstorage
- [x] **Outside community**) Re-login works without clearing localstorage

### Near

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage

### Phantom/Solana

- [x] **Inside community**) Login works before clearing localstorage
- [x] **Inside community**) Login works after clearing localstorage
- [x] **Inside community**) Logout works before clearing localstorage
- [x] **Inside community**) Re-login works without clearing localstorage
- [x] **Outside community**) Login works before clearing localstorage
- [x] **Outside community**) Login works after clearing localstorage
- [x] **Outside community**) Logout works before clearing localstorage
- [x] **Outside community**) Re-login works without clearing localstorage

### Polkadot/Substrate

- [x] **Inside community**) Login works before clearing localstorage
- [x] **Inside community**) Login works after clearing localstorage
- [x] **Inside community**) Logout works before clearing localstorage
- [x] **Inside community**) Re-login works without clearing localstorage
- [x] **Outside community**) Login works before clearing localstorage
- [x] **Outside community**) Login works after clearing localstorage
- [x] **Outside community**) Logout works before clearing localstorage
- [x] **Outside community**) Re-login works without clearing localstorage

### Ronin

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage

### Terra

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage

### Android: WalletConnect / desktop browser + 1inch mobile wallet

- [x] **Inside community**) Login works before clearing localstorage
- [x] **Inside community**) Login works after clearing localstorage
- [x] **Inside community**) Logout works before clearing localstorage
- [x] **Inside community**) Re-login works without clearing localstorage
- [x] **Outside community**) Login works before clearing localstorage
- [x] **Outside community**) Login works after clearing localstorage
- [x] **Outside community**) Logout works before clearing localstorage
- [x] **Outside community**) Re-login works without clearing localstorage

### Android: WalletConnect / mobile browser + 1inch mobile wallet

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage
- [ ] **Outside community**) Login works before clearing localstorage
- [ ] **Outside community**) Login works after clearing localstorage
- [ ] **Outside community**) Logout works before clearing localstorage
- [ ] **Outside community**) Re-login works without clearing localstorage

### Android: WalletConnect / desktop browser + keyring pro mobile wallet

- [x] **Inside community**) Login works before clearing localstorage
- [x] **Inside community**) Login works after clearing localstorage
- [x] **Inside community**) Logout works before clearing localstorage
- [x] **Inside community**) Re-login works without clearing localstorage
- [x] **Outside community**) Login works before clearing localstorage
- [x] **Outside community**) Login works after clearing localstorage
- [x] **Outside community**) Logout works before clearing localstorage
- [x] **Outside community**) Re-login works without clearing localstorage

### Android: WalletConnect / mobile browser + keyring pro mobile wallet

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage
- [ ] **Outside community**) Login works before clearing localstorage
- [ ] **Outside community**) Login works after clearing localstorage
- [ ] **Outside community**) Logout works before clearing localstorage
- [ ] **Outside community**) Re-login works without clearing localstorage

### Android: WalletConnect / desktop browser + now mobile wallet

- [x] **Inside community**) Login works before clearing localstorage
- [x] **Inside community**) Login works after clearing localstorage
- [x] **Inside community**) Logout works before clearing localstorage
- [x] **Inside community**) Re-login works without clearing localstorage
- [x] **Outside community**) Login works before clearing localstorage
- [x] **Outside community**) Login works after clearing localstorage
- [x] **Outside community**) Logout works before clearing localstorage
- [x] **Outside community**) Re-login works without clearing localstorage

### Android: WalletConnect / mobile browser + now mobile wallet

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage
- [ ] **Outside community**) Login works before clearing localstorage
- [ ] **Outside community**) Login works after clearing localstorage
- [ ] **Outside community**) Logout works before clearing localstorage
- [ ] **Outside community**) Re-login works without clearing localstorage

### iPhone: WalletConnect / desktop browser + 1inch mobile wallet

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage
- [ ] **Outside community**) Login works before clearing localstorage
- [ ] **Outside community**) Login works after clearing localstorage
- [ ] **Outside community**) Logout works before clearing localstorage
- [ ] **Outside community**) Re-login works without clearing localstorage

### iPhone: WalletConnect / mobile browser + 1inch mobile wallet

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage
- [ ] **Outside community**) Login works before clearing localstorage
- [ ] **Outside community**) Login works after clearing localstorage
- [ ] **Outside community**) Logout works before clearing localstorage
- [ ] **Outside community**) Re-login works without clearing localstorage

### iPhone: WalletConnect / desktop browser + keyring pro mobile wallet

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage
- [ ] **Outside community**) Login works before clearing localstorage
- [ ] **Outside community**) Login works after clearing localstorage
- [ ] **Outside community**) Logout works before clearing localstorage
- [ ] **Outside community**) Re-login works without clearing localstorage

### iPhone: WalletConnect / mobile browser + keyring pro mobile wallet

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage
- [ ] **Outside community**) Login works before clearing localstorage
- [ ] **Outside community**) Login works after clearing localstorage
- [ ] **Outside community**) Logout works before clearing localstorage
- [ ] **Outside community**) Re-login works without clearing localstorage

### iPhone: WalletConnect / desktop browser + now mobile wallet

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage
- [ ] **Outside community**) Login works before clearing localstorage
- [ ] **Outside community**) Login works after clearing localstorage
- [ ] **Outside community**) Logout works before clearing localstorage
- [ ] **Outside community**) Re-login works without clearing localstorage

### iPhone: WalletConnect / mobile browser + now mobile wallet

- [ ] **Inside community**) Login works before clearing localstorage
- [ ] **Inside community**) Login works after clearing localstorage
- [ ] **Inside community**) Logout works before clearing localstorage
- [ ] **Inside community**) Re-login works without clearing localstorage
- [ ] **Outside community**) Login works before clearing localstorage
- [ ] **Outside community**) Login works after clearing localstorage
- [ ] **Outside community**) Logout works before clearing localstorage
- [ ] **Outside community**) Re-login works without clearing localstorage

### CODE REVIEW

**Comment by mzparacha** [comment]
@jnaviask Just completed some tests. There are some race condition in state updates that are done in the login modal by the child components.

React state updates are async, so essentially all these callbacks that are setting value using `this.variable = ` here are not working correctly in the translated code
image


The solution here is to update the state in the functions directly and cleanup race conditions in the child components. 
@masvelio just wanted to get your eyes here as you did a similar thing with login modal sometime ago

**Comment by mzparacha** [comment]
Accidently merged the RQ research branch, reverting..

**Comment by mzparacha** [comment]
1inch, keyring and now-wallet not working on mobile-mobile for some reason, did some debugging nothing yet.. Will continue tomorrow

**Comment by masvelio** [comment]
> @jnaviask Just completed some tests. There are some race condition in state updates that are done in the login modal by the child components.
> 
> React state updates are async, so essentially all these callbacks that are setting value using `this.variable = ` here are not working correctly in the translated code image
> 
> The solution here is to update the state in the functions directly and cleanup race conditions in the child components. I also plan to lift the common code in login_desktop and login_mobile to the login_modal.
> 
> @masvelio just wanted to get your eyes here as you did a similar thing with login modal sometime ago

looks good now, incredible work on testing 💪  I remember that I skipped rewriting it to the react because we had to keeo going but I run into state problems and race conditions that would require a lot of refactor and testing. It seems that we have more time now and you do this job in right direction. Will be happy to take a look again before merge

**Comment by alexyoung23j** [comment]
Logins working on iphone for me.

**Review by jnaviask — CHANGES_REQUESTED** [review]
One small change to file structure, then will QA.

**Review by jnaviask — CHANGES_REQUESTED** [review]
One small change to file structure, then will QA.

File: packages/commonwealth/client/scripts/lib/wallet.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import IWebWallet from '../models/IWebWallet';
+import Near from '../controllers/chain/near/adapter';
```
[the remark this answers is not in the document]
Comment (mzparacha) [reply packages/commonwealth/client/scripts/lib/wallet.ts]: [Moved](https://github.com/hicommonwealth/commonwealth/pull/3714/commits/70588a571af4eae97f954ae7a6396103df265e56)

**Comment by jnaviask** [comment]
@mzparacha after completing login flow with a new account, modal does not close properly. Can get a loom if needed.

**Comment by mzparacha** [comment]
@jnaviask fixed the modal closing. As discussed it was a race condition that wasn't surfaced in the previous implementation.

When we create a new account, it calls the `initAppState` which triggers a re-render with `app.loginStateEmitter.emit('redraw');` which rerenders the app and the login modal is unmounted and remounted which then gives it a fresh state but the calling component of the login modal still holds the previous open state of the modal. Due to this the "welcome" dialog wasn't always showing after a new signup.

Fixed this race condition with a change to only redraw after the account signup is complete in cases of new accounts.

Ready is ready for review!

**Comment by mzparacha** [comment]
@jnaviask bump

File: packages/commonwealth/client/scripts/hooks/useWallets.tsx [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+      setPrimaryAccount(walletProps.initialAccount);
+      setAddress(walletProps.initialAccount.address);
+    }
+    // if (walletProps.initialWebWallet) {
+    //   this.selectedWallet = walletProps.initialWebWallet;
+    // }
+    if (walletProps.initialWallets) {
+      setWallets(walletProps.initialWallets);
+    }
+  }, []);
```
Comment (masvelio) [inline packages/commonwealth/client/scripts/hooks/useWallets.tsx:134]: what about missing dependencies?

File: packages/commonwealth/client/scripts/hooks/useWallets.tsx [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+        (state: boolean) => {
+          setShowMobile(state);
+        },
+        isWindowMediumSmallInclusive
+      )
+    );
+
+    return () => {
+      // eslint-disable-next-line no-restricted-globals
+      removeEventListener('resize', () =>
```
Comment (masvelio) [inline packages/commonwealth/client/scripts/hooks/useWallets.tsx:150]: can't we just use `window`?

File: packages/commonwealth/client/scripts/hooks/useWallets.tsx [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+    //   this.selectedWallet = walletProps.initialWebWallet;
+    // }
+    if (walletProps.initialWallets) {
+      setWallets(walletProps.initialWallets);
+    }
+  }, []);
+
+  useEffect(() => {
+    // eslint-disable-next-line no-restricted-globals
+    addEventListener('resize', () =>
```
Comment (masvelio) [inline packages/commonwealth/client/scripts/hooks/useWallets.tsx:138]: can't we just use `window`?

**Review by masvelio — APPROVED** [review]
Thanks for tidy things up and moving logic out of component to the custom hook! Left some small comments.

**Review by masvelio — APPROVED** [review]
Thanks for tidy things up and moving logic out of component to the custom hook! Left some small comments.

File: packages/commonwealth/client/scripts/hooks/useWallets.tsx (hunk 4 above)
Comment (mzparacha) [reply packages/commonwealth/client/scripts/hooks/useWallets.tsx]: Nice spot. This was actually already there and I tried to stay close to existing implementation as much as possible.

Currently, I added a new hook https://github.com/hicommonwealth/commonwealth/pull/3714/commits/2ba8e538596d67a2fd2d31f509961a43de0b8abe to easily hook to window resize events. There are a lot of places where we can just dry up code with it

Example:
```
useBrowserWindow({
    onResize: () => { /* resize function */ } 
    resizeListenerUpdateDeps: [... deps that on change will update the resize listener],
  });
```

File: packages/commonwealth/client/scripts/helpers/wallet.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+      return wallet.accounts[0].address;
+    }
+  })();
+
+  return selectedAddress;
+};
+
+const loginToAxie = async (loginUrl = '') => {
+  const result = await $.post(loginUrl, {
```
[the remark this answers is not in the document]
Comment (mzparacha) [reply packages/commonwealth/client/scripts/helpers/wallet.ts]: Done https://github.com/hicommonwealth/commonwealth/pull/3714/commits/98218e6a2a3f139c4230b32fe56578dba1bc54c2

**Review by jnaviask — CHANGES_REQUESTED** [review]
Logging in for the first time on a new account logs me in without showing the edit profile modal, and brings the modal back to the original state. Join state is also weird. See video: https://www.loom.com/share/12f7f8cdc9c34fb7a23fcd61e5932a82

**Comment by mzparacha** [comment]
@jnaviaske Not repro with metamask in dydx. Can you please try with clearing localstoarge 1 time after pulling and then regular testing?

https://github.com/hicommonwealth/commonwealth/assets/51641047/05a6ea08-c636-41ea-a00b-da2fd923ea25

**Comment by jnaviask** [comment]
@mzparacha same behavior here in an incognito tab -- maybe because the forced resize from my tiling WM (popping up the extension resizes the browser to half width) messes up the modal states?

**Comment by jnaviask** [comment]
@mzparacha Same issue does happen on current production -- okay, separate bug then.

**Comment by mzparacha** [comment]
@jnaviask bump for review

**Review by jnaviask — APPROVED** [review]
I think this is safe to merge, as the core flows work as well as they do on production, and we can catch edge case bugs in follow-up.

**Pull request merged**